### PR TITLE
Initial working book

### PR DESCRIPTION
**Pull request by EllenMooney95**
### Description
Create a function to subscribe to raw book data
This relates to issue #44 
Left to do: Remove price points where orderamount=0 

### Checklist

- [x] A short but descriptive title
- [x] A description that references the issue e.g. this pull request closes (#47).
- [x] Assignees responsible for moving the PR forward.
- [x] Colour-coded labels for categorisation.
- [x] A project to track PRs on a higher level

### CODE REVIEW

File: code/processes/bitfeed.q [hunk 1]
```diff
@@ -24,9 +24,10 @@ while[.bfx.notpconnected[];
 .bfx.tp:.servers.gethandlebytype[.bfx.tickerplanttypes;`any];
 
 .bfx.publish:{[t;x].bfx.tp(`.u.upd;t;flip x)};                                                         / Publish to tickerplant after replaylog
-.bfx.checkws:{@[{x .j.j `event`cid!(`ping;1234)};;{[x].bfx.init[]}] each .bfx`quote`trade};            / Send ping event which returns pong
+.bfx.checkws:{@[{x .j.j `event`cid!(`ping;1234)};;{[x].bfx.init[]}] each .bfx`quote`trade`bookh};            / Send ping event which returns pong
```
Comment (claire-sloan) [inline code/processes/bitfeed.q]: align comment

File: code/processes/bitfeed.q [hunk 2]
```diff
@@ -24,9 +24,10 @@ while[.bfx.notpconnected[];
...
 
-.bfx.data:();                                                                                          / Create empty list to insert data into
+.bfx.data:();                                                                                          / Create empty list to insert data intoi
+.bfx.book:()!();
```
Comment (claire-sloan) [inline code/processes/bitfeed.q]: add comment

File: code/processes/bitfeed.q [hunk 3]
```diff
@@ -44,10 +45,19 @@ while[.bfx.notpconnected[];
...
     ]
-  };
+  ]     
+ };
+
+.bfx.bookupd:{
+  $[99h~type a:.j.k x;                                                                                 / If incoming data is in dictionary format
+    if[`subscribed~`$a`event;
+      .bfx.bookchannels[a`chanId]:`$a`pair;                                                            / Check the event to make sure it is subscribe and not pong
+      .bfx.book[a`chanId]:1!flip `price`noorders`amount!()];
```
Comment (claire-sloan) [inline code/processes/bitfeed.q]: whitespace + comment lines in function

File: code/processes/bitfeed.q [hunk 4]
```diff
@@ -44,10 +45,19 @@ while[.bfx.notpconnected[];
...
+    if[not 10=type a 1;@[`.bfx.book;a 0;upsert;a 1]]                                                            / Change channel ID to sym from .bfx.channels
```
Comment (claire-sloan) [inline code/processes/bitfeed.q]: align comment

File: code/processes/bitfeed.q [hunk 5]
```diff
@@ -56,13 +66,14 @@ while[.bfx.notpconnected[];
...
   `.bfx.channels set ()!();                                                                            / Create empty dictionary to map syms to channel IDs
-  `.bfx.quotechannels set ()!();                                                                       / Create empty dictionary to map sym to quote channel ID     
+  `.bfx.quotechannels set ()!();                                                                       / Create empty dictionary to map sym to quote channel ID
+  `.bfx.bookchannels set ()!();  
   hclose each exec h from .ws.w;                                                                       / Make sure websockets are closed
   delete from `.ws.w;                                                                                  / Remove them from the websocket table
   .lg.o[`init;"Trying to open a websocket"];
-  (`.bfx.quote;`.bfx.trade) set' .ws.open[.bfx.url] each `.bfx.quoteupd`.bfx.tradeupd;                 / Open a websocket for both trades and quotes if necessary
+  (`.bfx.quote;`.bfx.trade;`.bfx.bookh) set' .ws.open[.bfx.url] each `.bfx.quoteupd`.bfx.tradeupd`.bfx.bookupd;                 / Open a websocket for both trades and quotes if necessary
```
Comment (claire-sloan) [inline code/processes/bitfeed.q]: whitespace

**Comment by jonathonmcmurray** [comment]
I'm not seeing anything to actually *build* the book; looks like you're just passing through the raw messages - do you guys understand what is meant by "building the book"? If not I can take you through the idea in the morning

**Comment by mkelly95** [comment]
The data that we get is a price level, number of orders and total amount at that level. It is the aggregated amount from all the orders so all that needs doing is upserting to the relevant keyed table and if the number of orders is 0, removed from that keyed table

**Comment by jonathonmcmurray** [comment]
@mkelly95 Not quite. That sounds like maintaining the *current* book in real time, but it's not how you would keep a record of the book over time, as is required here.

Essentially you should have a `book` table where each record is a snapshot of the full book at that moment in time (with a new record generated every time the book changes). So for example, could be a schema something like:

```
q)meta depth
c     | t f a
------| -----
time  | p
sym   | s   g
bid1  | f
bsize1| i
bid2  | f
bsize2| i
bid3  | f
bsize3| i
ask1  | f
asize1| i
ask2  | f
asize2| i
ask3  | f
asize3| i
```

(with number of levels configurable by user)

Alternatively, could have `bids`, `bsizes`, `asks` & `asizes` fields with nested lists in each to represent the full book. It's up to your team to determine which is best, and justify your decision when asked

File: code/processes/bitfeed.q [hunk 6]
```diff
@@ -44,10 +45,19 @@ while[.bfx.notpconnected[];
     if[`subscribed~`$a`event;.bfx.quotechannels[a`chanId]:`$a`pair];                                   / Check the event to make sure it is subscribe and not pong
     if[not 10=type a 1;
       a[0]:.bfx.quotechannels a 0;                                                                     / Change channel ID to sym from .bfx.channels
-      .bfx.publish[`quote_bfx]enlist raze[a] 0 1 2 3 4                                                 / Add sym, tradeID, size and price from data into .bfx.data
-      ]
+      .bfx.publish[`quote_bfx]enlist raze[a] 0 1 2 3 4                                                 / Add sym, tradeID, size and price from dx.data
     ]
-  };
+  ]
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: Semi colon.

File: code/processes/bitfeed.q [hunk 7]
```diff
@@ -44,10 +45,19 @@ while[.bfx.notpconnected[];
     if[`subscribed~`$a`event;.bfx.quotechannels[a`chanId]:`$a`pair];                                   / Check the event to make sure it is subscribe and not pong
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: I think this entire conditional could be cleaned up to
```
if[(99h~type a:.j.k x)&`subscribed~`$a`event;:.bfx.quotechannels...
if[not[99h~type a]&not 10=type 1;...
```

File: code/processes/bitfeed.q [hunk 8]
```diff
@@ -44,10 +45,19 @@ while[.bfx.notpconnected[];
...
-      .bfx.publish[`quote_bfx]enlist raze[a] 0 1 2 3 4                                                 / Add sym, tradeID, size and price from data into .bfx.data
-      ]
+      .bfx.publish[`quote_bfx]enlist raze[a] 0 1 2 3 4                                                 / Add sym, tradeID, size and price from dx.data
     ]
-  };
+  ]     
+ };
+
+.bfx.bookupd:{
+  $[99h~type a:.j.k x;                                                                                 / If incoming data is in dictionary format
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: Same as my comment on 45.

File: code/processes/bitfeed.q [hunk 9]
```diff
@@ -56,13 +66,14 @@ while[.bfx.notpconnected[];
 
 .bfx.init:{
   `.bfx.channels set ()!();                                                                            / Create empty dictionary to map syms to channel IDs
-  `.bfx.quotechannels set ()!();                                                                       / Create empty dictionary to map sym to quote channel ID     
+  `.bfx.quotechannels set ()!();                                                                       / Create empty dictionary to map sym to quote channel ID
+  `.bfx.bookchannels set ()!();
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: I think this should work:
```
.bfx[`quotechannels`bookchannels]:()!();
```

File: code/processes/bitfeed.q [hunk 10]
```diff
@@ -44,25 +45,40 @@ while[.bfx.notpconnected[];
...
-      .bfx.publish[`quote_bfx]enlist raze[a] 0 1 2 3 4                                                 / Add sym, tradeID, size and price from data into .bfx.data
-      ]
+      .bfx.publish[`quote_bfx]enlist raze[a] 0 1 2 3 4                                                 / Add sym, tradeID, size and price from dx.data
     ]
-  };
+  ]     
+ };
+
+.bfx.bookupd:{
+    $[99h~type a:.j.k x;                                                                               / If data type is dictionary
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: This whole conditional is a nightmare to read.

File: code/processes/bitfeed.q [hunk 11]
```diff
@@ -44,26 +47,43 @@ while[.bfx.notpconnected[];
...
+  ]     
+ };
+
+.bfx.bookupd:{
+    $[99h~type a:.j.k x;                                                                               / If data type is dictionary             
+      if[`subscribed~`$a`event;                                                                        / If the event is subscribed
+        .bfx.bookchannels[a`chanId]:`$a`pair;                                                          / Add channelId and pair to dict
+        .bfx.book[a`chanId]:1!flip `price`noorders`amount!()];                                         / Add empty book for chanbelId
+    10=type a 1;                                                                                       / Else, if is heartbeat record it
+    `.bfx.lasthb set .z.p;
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: Why are these actions no longer indented? Very unclear.

File: code/processes/bitfeed.q [hunk 12]
```diff
@@ -24,9 +24,14 @@ while[.bfx.notpconnected[];
...
-.bfx.checkws:{@[{x .j.j `event`cid!(`ping;1234)};;{[x].bfx.init[]}] each .bfx`quote`trade};            / Send ping event which returns pong
+.bfx.checkws:{@[{x .j.j `event`cid!(`ping;1234)};;{[x].bfx.init[]}] each .bfx`quote`trade`bookh};      / Send ping event which returns pong
+.bfx.checkhb:{if[00:00:02<t:.z.p - .bfx.lasthb;.lg.o[`inf;"last hearbeat was ",string[t],"s ago"]]};
 
-.bfx.data:();                                                                                          / Create empty list to insert data into
+.bfx.data:();                                                                                          / Create empty list to insert data intoi
+.bfx.book:()!();
+.bfx.lasthb:0Np;
+.bfx.depth:`25`100 25<first"I"$.proc.params[`depth];
+.bfx.cdepth:first"I"$.proc.params[`depth];
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: No squares needed for here + 33 but may need some comments.
Also this line and 33 share some duplicate code, how about this instead:
```
.bfx.depth:`25`100 25<.bfx.cdepth:first"I"$.proc.params`depth;
```

File: code/processes/bitfeed.q [hunk 13]
```diff
@@ -44,26 +49,45 @@ while[.bfx.notpconnected[];
...
+        .bfx.book[a`chanId]:`bid`ask!2#enlist 1!flip `price`noorders`amount!()];                       / Add empty book for chanbelId
+    10=type a 1;                                                                                       / Else, if is heartbeat record it
+      `.bfx.lasthb set .z.p;
+    9=type a[1;1];
+      {.[`.bfx.book;(x 0;y);upsert;z]}[a]'[`bid`ask;(0;.bfx.depth) cut a 1];                           / Upsert quote to book
+    0f=a[1;1];
+      .[`.bfx.book;;_;a[1;0]] a[0],`ask`bid -1 1f ? a[1;2];                                            / If no.orders=0 remove records with 0 orders
+      {.[`.bfx.book;(x 0;y);upsert;x 1]}[a] `ask`bid a[1;2]>0
+  ];
+   if[not 99h~type a;
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: The previous conditional will also perform actions based on this condition, maybe this needs merge with above or lines 62-69 placed in here.

File: code/processes/bitfeed.q [hunk 14]
```diff
@@ -44,26 +49,45 @@ while[.bfx.notpconnected[];
...
+    .bfx.publish[`book_bfx] enlist .bfx.bookchannels[a 0],0!'.bfx.cdepth#'((xasc;xdesc)@'`price)@'.bfx.book[8173f]`ask`bid]
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: I think you should avoid sorting twice with `xasc` + `xdesc` and instead sort once.
Something like:
```
0!'(1 -1*.bfx.cdepth)#\:`price xasc/:.bfx.book[8173f]`ask`bid
```
Just need to ensure the `xasc` runs on both tables there.

File: code/processes/bitfeed.q [hunk 15]
```diff
@@ -24,9 +24,13 @@ while[.bfx.notpconnected[];
 .bfx.tp:.servers.gethandlebytype[.bfx.tickerplanttypes;`any];
 
 .bfx.publish:{[t;x].bfx.tp(`.u.upd;t;flip x)};                                                         / Publish to tickerplant after replaylog
-.bfx.checkws:{@[{x .j.j `event`cid!(`ping;1234)};;{[x].bfx.init[]}] each .bfx`quote`trade};            / Send ping event which returns pong
+.bfx.checkws:{@[{x .j.j `event`cid!(`ping;1234)};;{[x].bfx.init[]}] each .bfx`quote`trade`bookh};      / Send ping event which returns pong
+.bfx.checkhb:{if[00:00:02<t:.z.p - .bfx.lasthb;.lg.o[`inf;"last hearbeat was ",string[t],"s ago"]]};
 
-.bfx.data:();                                                                                          / Create empty list to insert data into
+.bfx.data:();                                                                                          / Create empty list to insert data intoi
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: Typo in them comment.

**Pull request merged**